src/ProtocolZoo/ProtocolZoo.jl

```diff
@@ -166,8 +170,19 @@ end
     rounds = prot.rounds
     round = 1
     while rounds != 0
-        a = findfreeslot(prot.net[prot.nodeA], randomize=prot.randomize)
-        b = findfreeslot(prot.net[prot.nodeB], randomize=prot.randomize)
+        freeA = sum([!isassigned(prot.net[prot.nodeA][i]) for i in 1:length(prot.net[prot.nodeA].staterefs)])
+        freeB = sum([!isassigned(prot.net[prot.nodeB][i]) for i in 1:length(prot.net[prot.nodeB].staterefs)])
+        isentangled = !isnothing(query(prot.net[prot.nodeA], EntanglementCounterpart, prot.nodeB, ❓;assigned=true))
+        margin = isentangled ? prot.margin : prot.hardmargin
+        if freeA >= prot.margin && freeB >= prot.margin
+            a = findfreeslot(prot.net[prot.nodeA]; randomize=prot.randomize)
+            b = findfreeslot(prot.net[prot.nodeB]; randomize=prot.randomize)
```

Comment:
This seems a bit roundabout. `freeA` and `freeB` basically are a reimplementation of `findfreeslot`. We should have something a bit more [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)

Let's have it instead more in the following style:

```
isentangled = !isnothing(query...)
margin = isentangled ? ...
a = findfreeslots(...; ..., margin)
b = findfreeslots(...; ..., margin)
```

That was we avoid reimplementing findfreeslot and have this capability available to other protocols. I believe this is the important abstraction to make available.